lib/Dialect/FIRRTL/Transforms/LowerOpenAggs.cpp

```diff
@@ -0,0 +1,629 @@
+        if (pmi.identity) {
+          // (Just do the RAUW here instead of tracking the mapping for this
+          // too.)
+          assert(oldResult.getType() == newInst.getType(newIndex));
+          oldResult.replaceAllUsesWith(newInst.getResult(newIndex));
+          return success();
+        }
+
+        // Create mappings for updating open aggregate users.
+        auto newPortIndex = newIndex;
+        if (pmi.hwType)
+          hwOnlyAggMap[oldResult] = newInst.getResult(newPortIndex++);
+        else
+          hwOnlyAggMap[oldResult] = std::nullopt;
+
+        for (auto &field : pmi.fields) {
+          auto ref = FieldRef(oldResult, field.fieldID);
+          auto newVal = newInst.getResult(newPortIndex++);
+          assert(newVal.getType() == field.type);
+          nonHWValues[ref] = newVal;
+        }
+        return success();
+      });
+  if (failed(mappingResult))
+    return failure();
+
+  opsToErase.push_back(op);
+
+  return success();
+}
+
+//===----------------------------------------------------------------------===//
+// Type Conversion
+//===----------------------------------------------------------------------===//
+
+PortMappingInfo Visitor::mapPortType(Type type) {
+  PortMappingInfo pi{false, {}, {}};
+  auto ftype = dyn_cast<FIRRTLType>(type);
+  // Ports that aren't open aggregates are left alone.
+  if (!ftype || !isa<OpenBundleType, OpenVectorType>(ftype)) {
+    pi.identity = true;
+    return pi;
+  }
+
+  // TODO: Manage SmallString for suffix instead of Twine of pieces (since will
+  // conv to vector anyway).
+  // NOLINTBEGIN(misc-no-recursion)
+  auto recurse = [&](auto &&f, FIRRTLType type, const Twine &suffix = "",
+                     bool flip = false,
+                     uint64_t fieldID = 0) -> FIRRTLBaseType {
+    return TypeSwitch<FIRRTLType, FIRRTLBaseType>(type)
+        .Case<FIRRTLBaseType>([](auto base) { return base; })
+        .template Case<OpenBundleType>(
+            [&](OpenBundleType obTy) -> FIRRTLBaseType {
+              SmallVector<BundleType::BundleElement> hwElements;
+              for (const auto &[index, element] :
+                   llvm::enumerate(obTy.getElements()))
+                if (auto base =
+                        f(f, element.type, suffix + "_" + element.name.strref(),
+                          flip ^ element.isFlip,
+                          fieldID + obTy.getFieldID(index)))
+                  hwElements.emplace_back(element.name, element.isFlip, base);
+
+              if (hwElements.empty())
+                return FIRRTLBaseType{};
+
+              return BundleType::get(context, hwElements, obTy.isConst());
+            })
+        .template Case<OpenVectorType>(
+            [&](OpenVectorType ovTy) -> FIRRTLBaseType {
+              FIRRTLBaseType convert;
+              for (auto idx : llvm::seq<size_t>(0U, ovTy.getNumElements()))
+                convert = f(f, ovTy.getElementType(), suffix + "_" + Twine(idx),
```

Comment:
This both converts and enumerates the non-hw leaf elements-- which need to be tracked per-element of the vector (each non-hw leaf becomes a new port).

The reason `convert` is rewritten is because of what you say, only need the resulting type once (but need to walk for each index).  This could certainly be clearer, I'll touch this up, thanks!

I.. expect this could be done more efficiently, esp re:leveraging available reuse opportunities (mapping for same type globally, signatures of modules match their instances, so on).  Root-relative suffixes for the mappings make caching the nested type here not trivial, but that's just one detail re:current approach.
I do not want this to show up on a profile, if you see a problem, missed opportunity, or have sense this is going to be a problem LMK.   Either way if this shows up on a profile I'd love a reason to engineer this better :+1: .

Reply:
I see, the clarifying comment and assertion definitely help